File: in_Python/0811 Subdomain Visit Count.py
class Solution(object):
    def subdomainVisits(self, cpdomains):
        """
        :type cpdomains: List[str]
        :rtype: List[str]
        """
        res = []
        dict = {}
        for x in cpdomains:
            t = x.split(' ')
            count = int(t[0])
            dnames = t[1].split('.')
            if len(dnames) == 3:
                s1 = dnames[1] + '.' + dnames[2]
                s2 = dnames[2]
                s3 = t[1]
                if s1 not in dict:
                    dict[s1] = count
                else:
                    dict[s1] += count
                if s2 not in dict:
                    dict[s2] = count
                else:
                    dict[s2] += count
                if s3 not in dict:
                    dict[s3] = count
                else:
                    dict[s3] += count
            else:
                s1 = dnames[1]
                s2 = t[1]
                if s1 not in dict:
                    dict[s1] = count
                else:
                    dict[s1] += count
                if s2 not in dict:
                    dict[s2] = count
                else:
                    dict[s2] += count                
        for k in dict:
            res.append(str(dict[k]) + ' ' + k)
        return res
                
        

File: in_Java/0811 Subdomain Visit Count.java
class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        List<String> res = new ArrayList<>();
        if(cpdomains == null || cpdomains.length == 0) return res;
        HashMap<String, Integer> hm = new HashMap<>();
        for(String s : cpdomains){
            String[] ss = s.split(" ");
            String x = ss[1];
            int a = Integer.parseInt(ss[0]);
            List<Integer> t = new ArrayList<>();
            t.add(-1);
            for(int i = 0; i < x.length(); i++){
                if(x.charAt(i) == '.') t.add(i);
            }
            for(int i : t){
                String cur = x.substring(i+1, x.length());
                if(!hm.containsKey(cur)) hm.put(cur, a);
                else hm.put(cur, hm.get(cur)+a);
            }
        }
        for(String k : hm.keySet()){
            res.add(hm.get(k) + " " + k);
        }
        return res;
    }
}